src/graph.ts

```diff
@@ -1,30 +1,88 @@
-import { PackageInfos } from "./types/PackageInfo";
-import { getInternalDeps } from "./dependencies";
+import type { PackageInfos } from "./types/PackageInfo";
+import type { DependencyMap } from "./createDependencyMap";
+import type { PackageGraph, PackageGraphVisitor } from "./types/PackageGraph";
 
-const graphCache = new Map<PackageInfos, [string, string][]>();
+import { createDependencyMap } from "./createDependencyMap";
+import multimatch from "multimatch";
 
-export function getPackageGraph(packages: PackageInfos) {
-  if (graphCache.has(packages)) {
-    return graphCache.get(packages)!;
-  }
+// Reference: https://github.com/pnpm/pnpm/blob/597047fc056dd25b83638a9ab3df0df1c555ee49/packages/filter-workspace-packages/src/parsePackageSelector.ts
+export interface PackageGraphScope {
+  namePatterns?: string[];
+  includeDependencies?: boolean;
+  includeDependents?: boolean;
+}
```

Comment:
We should modify this type... :)

Add
- includeExternalDependencies?: boolean

Future work:
- includeDevDependencies
- includePeerDependencies
(To include dev and peer we should add metadata on the edges to know what sort of relationship it is, and probably the version as well, this is probably best leaving out of scope for now)

Reply:
Does it make sense to have includeDependencies and includeDependents if there is no name pattern specified? 

There are two things that we are going for: the complete graph and the subgraphs. I think for the complete graph, the scope should not be given at all, and for the subgraphs, a name pattern should be _required_ for the scope.

Reply:
and if we are to keep includeDependencies and includeDependents as options when there is no name pattern given, too,

if A and B depend on C,
would we expect includeDependents=true, includeDependencies = false to return A and B only? and includeDependencies = true, includeDependents= false to return C only?

Right now, it lists A, B, and C as packages for both of the cases above, and also contains all of the edges.

Reply:
another thing is: when includeDependencies and includeDependents are both set to false, we only get the namePatterns we enter back as the package names, and this is  not useful. I don't think it makes sense to have the option to set both of them to false. Maybe getting dependencies, dependents, or all affected should be the options.